### ISSUE TITLE
processing_metadata structure

### ISSUE THREAD

**Issue by dwief-usgs** [usgs-bcb/bispy#11]
Posting for discussion.  In the use cases of species database development we are often using classes/modules out of bispy to return information based on a "target", which in this case is a species name.  At the end of the exercise we want to join all of this information back together.  I suggest we add key to represent the target at hand.

An example to show where this might be helpful.  I have a species name Fagus grandifolia that I want to search ScienceBase for Data Releases on common names of this specie using bispy.sb.Search.search_snapshot.  If you pass the common name -> American Beech Tree as q there is nothing that documents this item as being a search for Fagus grandifolia.  When we want to grab all data on this species from our cached processes we need to run additional code to reconstruct connections between Sci Name and common names.

**Comment by skybristol** [comment]
I agree that this needs some new thinking. Currently, I stored the scientific name that was the subject of a search in a separate top level key on each document called "parameters." But I know that there has been a little variation in what got stored here and how as I was working through the data models for each thing in the cache.

Providing a key for each document in a cache is probably a good idea, but there are lots of ways of dealing with it. We are finally making some progress on implementing PROV in the context of data processing workflows like this running as Lambdas in the cloud, so we should probably align efforts here with that overall approach. In the case of a standalone workflow producing a point in time data cache rather than a streaming workflow operating through time to refresh data in a persistent data store, the PROV trace could be incorporated directly into each resulting data entity like we are essentially doing now, or we could still externalize it into a related store. I'm a little more of a fan of the former approach, essentially replacing the processing_metadata and parameters structures we have now with a new "provenance" data structure containing the prov trace for the cached document.

I'll try to put some thinking into this in the next week or two and will touch base with the FORT guy who is looking into this for the USNVC and SGCN use cases they are implementing on AWS.
